src/main/java/com/redhat/devtools/lsp4ij/features/documentSymbol/LSPDocumentSymbolBreadcrumbsInfoProvider.java

```diff
@@ -0,0 +1,152 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ *
+ * Contributors:
+ *     Red Hat Inc. - initial API and implementation
+ *******************************************************************************/
+
+package com.redhat.devtools.lsp4ij.features.documentSymbol;
+
+import com.intellij.ide.structureView.StructureViewTreeElement;
+import com.intellij.lang.Language;
+import com.intellij.lang.LanguageUtil;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.Conditions;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.ui.breadcrumbs.BreadcrumbsProvider;
+import com.intellij.util.containers.ContainerUtil;
+import com.redhat.devtools.lsp4ij.LanguageServiceAccessor;
+import com.redhat.devtools.lsp4ij.client.features.EditorBehaviorFeature;
+import com.redhat.devtools.lsp4ij.features.documentSymbol.LSPDocumentSymbolStructureViewModel.LSPDocumentSymbolViewElement;
+import com.redhat.devtools.lsp4ij.features.documentSymbol.LSPDocumentSymbolStructureViewModel.LSPFileStructureViewElement;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.Icon;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Document symbol-based breadcrumbs info provider.
+ */
+public class LSPDocumentSymbolBreadcrumbsInfoProvider implements BreadcrumbsProvider {
+
+    @Override
+    public Language[] getLanguages() {
+        // Register for all languages and filter in isSupported()
+        List<Language> languages = LanguageUtil.getLanguages(Conditions.alwaysTrue());
+        return languages.toArray(Language.EMPTY_ARRAY);
+    }
+
+    private boolean isSupported(@NotNull PsiElement element) {
+        Project project = element.getProject();
+        PsiFile file = element.getContainingFile();
+        VirtualFile virtualFile = file.getVirtualFile();
+        return LanguageServiceAccessor.getInstance(project).hasAny(
+                virtualFile,
+                ls ->
+                        // The document symbols feature must be available for the file's language server
+                        ls.getClientFeatures().getDocumentSymbolFeature().isEnabled(file) &&
+                                ls.getClientFeatures().getDocumentSymbolFeature().isSupported(file) &&
+                                // And this feature must be enabled
+                                EditorBehaviorFeature.enableDocumentSymbolsBreadcrumbsInfoProvider(file)
+        );
+    }
+
+    @Override
+    public boolean acceptElement(@NotNull PsiElement element) {
+        // NOTE: This is only used to determine which already-vetted elements should be included in the breadcrumb trail
+        return element instanceof DocumentSymbolData;
+    }
+
+    @Override
+    @Nullable
+    public PsiElement getParent(@NotNull PsiElement element) {
```

Comment:
The following two methods are how the caller determines what to show based on the offset. I wish it passed the offset itself, but instead it passes the element at that offset. In a plain text/TextMate LSP4IJ file, we do now have PSI elements, but only for semantic tokens. The intermediate space does not produce a distinct element aside from the file itself. That's why we're going to need to use inheritable thread locals with the original requested offset. More on that later, but just setting the context.